libraries/chain/chain_controller.cpp

```diff
@@ -490,6 +497,132 @@ void chain_controller::apply_block(const signed_block& next_block, uint32_t skip
    });
 }
 
+template<typename T>
+void check_output(const T& expected, const T& actual) {
+   EOS_ASSERT((expected == actual), block_tx_output_exception, 
+      "Value mismatch, expected: ${expected}, actual: ${actual}", ("expected", expected)("actual", actual));
+}
+
+template<typename T>
+void check_output(const vector<T>& expected, const vector<T>& actual) {
+   EOS_ASSERT((expected.size() == actual.size()), block_tx_output_exception,
+      "Vector size mismatch, expected: ${expected}, actual: ${actual}", ("expected", expected.size())("actual", actual.size()));
+   
+   for(int idx=0; idx < expected.size(); idx++) {
+      const auto &expected_element = expected.at(idx);
+      const auto &actual_element = actual.at(idx);
+      try {
+        check_output(expected_element, actual_element);
+      } FC_RETHROW_EXCEPTIONS( warn, "at index ${idx}", ("idx", idx ) );
+   }
+}
+
+template<>
+void check_output(const types::Bytes& expected, const types::Bytes& actual) {
+   EOS_ASSERT((expected.size() == actual.size()), block_tx_output_exception,
+      "Binary blob size mismatch, expected: ${expected}, actual: ${actual}", ("expected", expected.size())("actual", actual.size()));
+
+   EOS_ASSERT((std::memcmp(expected.data(), actual.data(), expected.size()) == 0), block_tx_output_exception,
+      "Binary blob contents differ");   
+}
+
+template<>
+void check_output(const types::AccountPermission& expected, const types::AccountPermission& actual) {
+   try {
+      check_output(expected.account, actual.account);
+   } FC_RETHROW_EXCEPTIONS(warn, "in .account");
+   try {
+      check_output(expected.permission, actual.permission);
+   } FC_RETHROW_EXCEPTIONS(warn, "in .permission");
+}
+
+template<>
+void check_output(const types::Message& expected, const types::Message& actual) {
+   try {
+      check_output(expected.code, actual.code);
```

Comment:
why don't you move the try{ } catch{} . into check_output and pass the "in .x" as a const char* 3rd argument?  I think it would make the code more readable.